sysvabi64/sysvabi64.rst

```diff
@@ -916,10 +923,639 @@ These follow the AAELF64_ base definition from Arm.
 
    PageBreak oneColumn
 
-Program Linking and Dynamic Linking
+Program Loading and Dynamic Linking
 ===================================
 
-Not available for this Alpha release.
+The system works on creating a program/process image by logically
+copying or mapping the program image from storage into a virtual
+memory segment. In most systems with demand paging, when the actual
+page is mapped into physical memory depends on the execution behaviour
+of the program and the working set that it requires for its correct
+execution. To minimize the number of such reads from the physical
+medium containing the program, executables and shared objects must
+have program segments whose file offsets and virtual addresses are
+congruent modulo the page size.
+
+Since the architecture supports a maximum page size of 64KiB the page
+size should be set to 64KiB for compatibility with systems that use
+64KiB pages. There are a number of optimizations that can be
+applied to save space.
+
+ * The first page of text contains the ELF header, program header
+   information and other information from the ELF file.
+
+ * The last page of text contains parts of the first page of data.
+
+ * The first data page has a copy of the last page of text.
+
```

Comment:
This bit always makes my head hurt!

The ELF file needs to satisfy p_vaddr is congruent to `p_offset` (modulo `p_align`). The linker can round both the `p_offset` and `p_vaddr` up to the next `p_align` boundary, putting zeroes in the file or it can add `p_align` bytes to `p_addr` with no padding needed in the file.
  
The best explanation I've read is the commit message for ld.bfd PR28824  https://sourceware.org/bugzilla/show_bug.cgi?id=28824#c3

In summary:
There will be a minimum of 2 pages required, one for the RO and one for the RW. In many cases mapping the same part of the file twice results in a smaller file, which could be significant for a really high page size such as 1 MiB on X86_64. In some cases where the RO almost fills up a full page there isn't much saving to be made by mapping the same page of the file twice. The `DATA_SEGMENT_ALIGN` linker script command has a calculation that pads the end of the RO in the file.

There is then the difference between max page size and common page size, that also makes my head hurt. The best explanation of this is in https://reviews.llvm.org/D33630?id=102280#767301